### DISCUSSION TITLE
Σ & ∏

### DISCUSSION THREAD

**Discussion by yukulele** [josdejong/mathjs#3075] [Ideas]
I would like to be able to use the sum (Σ) or the product (∏)

like this
```
 10
  Σ  2⁻ⁿ cos(2ⁿ π x)
 k=0
```

Unfortunately `math.sum()` only allow to sum parameters or matrix terms.

I suggest to add a new signature for `sum` (and `prod`) if 3rd term is a function:
```javascript
sum(initVal, endVal, callback)
```

example:
```js
sum(0, 10, f(n) = 2^-n cos(2^n pi x))
```

**Comment by josdejong** [comment]
Thanks for your suggestion, it's an interesting idea. 

Maybe it would be better to have the function as first argument:

```
sum(fn, start, end)
```
Because that is how you would read it aloud: "The sum of ... for k is 0 to 10". 

Note that this function is just eye candy for:

```js
sum(range(initVal, endVal)).map(callback)
```
So I'm not entirely sure it would add much. It reads very natural though, and with `toTex` we could output the correct math syntax for this expression.

EDIT: updated the example

**Reply by yukulele** [reply]
> Note that this function is just eye candy for:
> ```js
> range(initVal, endVal).map(callback)
> ```

I guess it's more like this:
```
sum(range(initVal, endVal).map(callback))
```

It does the job! ...maybe a little verbose?

---

> Maybe it would be better to have the function as first argument

not sure... 
- visually, bound are place before expression in the Σ notation
- and it must be hard to read if the function is long:
```js
sum(f(x) = 1^n * cos(3^n * pi * x), 0, 10)
```

a more explicit syntax might be:
```js
sum(start to end, fn)
sum(0 to 10, f(x) = x^2)
```

**Reply by josdejong** [reply]
Ah, you're right, I forgot the `sum(...)` around `range()`, updated that now.

I guess the order is personal preference, I personally do not find it hard to read but that may be just me (knowing what to expect). It's not a strong preference though. I'm curious to hear what others like best. And maybe we can have a look at other math application to see what syntax they use.
